File: knot/compile/constraints.py
"""Constraint validation emitter — turn each ``Constraint.body`` (an
``Expr`` from the semantic builder) into a SELECT that surfaces rows
of the primary class which violate the predicate.

Each emitted SELECT has the uniform shape:

    SELECT
        '<rule_id>'    AS rule_id,
        '<class_name>' AS class_name,
        '<severity>'   AS severity,
        <message>      AS message,
        <pk_col>       AS offending_pk
    FROM <schema>.<class><layer>
    WHERE NOT (<body_sql>);

Empty result → constraint passes. Non-empty rows are violations; the host
inspects ``severity`` to decide block-vs-warn.

``layer`` defaults to ``Layer.RESOLVED`` so validation runs against
the resolver's per-slot argmax view. Pass ``Layer.CANONICAL`` to target
the canonical table directly — appropriate only for write-direct
workflows where the host writes the canonical table itself.

``compile_sql(body, schema, layer)`` does the qualification —
no AST rewriting needed here, because the builder produces references
keyed by class name + slot name that render with the same suffix used
in the wrapping FROM clause.
"""

from __future__ import annotations

from collections.abc import Iterator

from knot.ast.select import Layer
from knot.ast.types import ClassRef
from knot.compile.expr import compile_sql
from knot.spec import ClassKind, Constraint, OntologyClass, Severity, Spec


def _escape_literal(s: str) -> str:
    """Escape a string for a single-quoted SQL literal."""
    return s.replace("'", "''")


def builtin_constraints(spec: Spec) -> Iterator[Constraint]:
    """Yield invariants knot ships automatically from the spec's shape.

    Two families today, both prefixed ``_builtin_`` so hosts can
    filter them out by name if they want:

    1. ``_builtin_fk_orphan_<Class>_<slot>`` — every ClassRef slot.
       Body: ``slot.is_null() | slot.target_exists()`` — valid rows
       are those where the FK is null (not provided) or points at a
       real canonical_id in the target's resolved view. ER's forward
       FK translation should keep this satisfied; this constraint
       catches the case where it didn't (e.g. a referenced binding
       was never ER-stamped).
    2. ``_builtin_required_null_<Class>_<slot>`` — every required
       non-identifier slot. Body: ``slot.is_not_null()`` — catches
       the case where every source's claim for a required slot was
       null, leaving the resolved row with NULL.

    Severity defaults to ERROR for FK orphans (structural — should
    never fail post-ER) and WARNING for required nulls (sources are
    expected to fill required slots but the host may want to
    tolerate gaps during early ingest).
    """
    for cls in spec.classes.values():
        # Built-ins only apply to concrete OntologyClass — virtual
        # classes inherit constraints from their concrete root.
        if not isinstance(cls, OntologyClass):
            continue
        if cls.kind != ClassKind.CONCRETE:
            continue
        for slot in cls.effective_slots():
            if isinstance(slot.type, ClassRef):
                fk_ref = cls.col[slot.name]
                yield Constraint(
                    name=f"_builtin_fk_orphan_{cls.name}_{slot.name}",
                    primary=cls,
                    body=fk_ref.is_null() | fk_ref.target_exists(),
                    severity=Severity.ERROR,
                    message=(
                        f"{cls.name}.{slot.name} → {slot.type.target.name}: "
                        f"FK value does not match any canonical_id"
                    ),
                )
            if slot.required and not slot.identifier:
                yield Constraint(
                    name=f"_builtin_required_null_{cls.name}_{slot.name}",
                    primary=cls,
                    body=cls.col[slot.name].is_not_null(),
                    severity=Severity.WARNING,
                    message=(
                        f"required slot {cls.name}.{slot.name} resolved to NULL "
                        f"(no source provided a non-null value)"
                    ),
                )


def emit_validation(
    spec: Spec,
    *,
    schema: str = "knot_data",
    layer: Layer = Layer.RESOLVED,
    scope_to_source_identifiers: dict[str, list[str]] | None = None,
    include_builtins: bool = True,
) -> list[tuple[str, str]]:
    """Return ``(constraint_name, validation_sql)`` pairs.

    Each ``validation_sql`` returns zero rows when the constraint holds
    and one row per violating canonical_id otherwise.

    ``include_builtins`` (default ``True``): also emit invariants
    knot derives from the spec shape — FK-orphan checks per
    ClassRef slot, required-slot-null checks per required slot.
    See ``builtin_constraints`` for details. Pass ``False`` to get
    only the user-declared constraints.

    ``scope_to_source_identifiers`` — when provided, each validation
    SELECT is restricted to the canonical_ids touched by that batch.
    The dict maps ``source_name`` → list of ``source_identifier`` values.
    Inlines the (source, identifier) tuples as SQL literals — no
    parameter list (same posture as ``Query.sql``). ``None`` or empty
    dict means no scoping (full-table validation, the default).
    """
    # Build the optional IN-subquery fragment once (shared across all constraints).
    scope_sql: str | None = None
    if scope_to_source_identifiers:
        pairs: list[str] = []
        for src, idents in scope_to_source_identifiers.items():
            src_lit = "'" + src.replace("'", "''") + "'"
            for ident in idents:
                ident_lit = "'" + ident.replace("'", "''") + "'"
                pairs.append(f"({src_lit}, {ident_lit})")
        if pairs:
            pairs_sql = ", ".join(pairs)
            scope_sql = pairs_sql  # stored; injected per-constraint below

    all_constraints = list(spec.constraints)
    if include_builtins:
        all_constraints.extend(builtin_constraints(spec))

    out: list[tuple[str, str]] = []
    for c in all_constraints:
        primary = c.primary
        identifier = primary.identifier_slot()
        table = f"{schema}.{primary.name.lower()}{layer}"
        bindings_table = f"{schema}.{primary.name.lower()}_bindings"
        message_literal = f"'{_escape_literal(c.message)}'" if c.message else "NULL"
        # outer_class threads the constraint's primary class through so
        # correlated ``this.<Primary>`` refs inside Aggregate predicates
        # bind to the FROM <primary> row, not the subquery's row.
        body_sql = compile_sql(
            c.body, schema=schema, layer=layer, outer_class=primary.name
        )
        scope_clause = ""
        if scope_sql is not None:
            scope_clause = (
                f"\n  AND {table}.{identifier.name} IN (\n"
                f"    SELECT DISTINCT canonical_id FROM {bindings_table}\n"
                f"    WHERE (source_name, source_identifier) IN ({scope_sql})\n"
                f"      AND canonical_id IS NOT NULL\n"
                f"  )"
            )
        sql = (
            f"SELECT\n"
            f"    '{_escape_literal(c.name)}' AS rule_id,\n"
            f"    '{_escape_literal(primary.name)}' AS class_name,\n"
            f"    '{_escape_literal(c.severity)}' AS severity,\n"
            f"    {message_literal} AS message,\n"
            f"    {identifier.name} AS offending_pk\n"
            f"FROM {table}\n"
            f"WHERE NOT ({body_sql}){scope_clause};"
        )
        out.append((c.name, sql))
    return out


def emit_validation_union(
    spec: Spec,
    *,
    schema: str = "knot_data",
    layer: Layer = Layer.RESOLVED,
    scope_to_source_identifiers: dict[str, list[str]] | None = None,
    include_builtins: bool = True,
) -> str | None:
    """Return a single ``UNION ALL`` of every constraint's validation
    SELECT, or ``None`` if the spec has no constraints."""
    parts = [
        sql.rstrip(";")
        for _, sql in emit_validation(
            spec,
            schema=schema,
            layer=layer,
            scope_to_source_identifiers=scope_to_source_identifiers,
            include_builtins=include_builtins,
        )
    ]
    if not parts:
        return None
    return "\nUNION ALL\n".join(parts) + ";"

File: src/main/java/knot/compile/Constraints.java
package knot.compile;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import knot.ast.expr.FkRef;
import knot.ast.expr.IsNull;
import knot.ast.select.Layer;
import knot.ast.types.ClassRef;
import knot.spec.ClassKind;
import knot.spec.Constraint;
import knot.spec.OntologyClass;
import knot.spec.Severity;
import knot.spec.Slot;
import knot.spec.Spec;

/**
 * Constraint validation SELECT emitter.
 *
 * <p>Turns each {@link Constraint#body()} (an {@code Expr} from the semantic builder) into a
 * SELECT that surfaces rows of the primary class which violate the predicate.
 *
 * <p>Each emitted SELECT has the uniform 5-column shape:
 * <pre>
 *   SELECT
 *       '<rule_id>'    AS rule_id,
 *       '<class_name>' AS class_name,
 *       '<severity>'   AS severity,
 *       <message>      AS message,
 *       <pk_col>       AS offending_pk
 *   FROM <schema>.<class><layer>
 *   WHERE NOT (<body_sql>);
 * </pre>
 *
 * <p>Empty result → constraint passes. Non-empty rows are violations; the host inspects
 * {@code severity} to decide block-vs-warn.
 *
 * <p>Ports {@code knot/compile/constraints.py}.
 */
public final class Constraints {

    private Constraints() {}

    // -------------------------------------------------------------------------
    // Public records
    // -------------------------------------------------------------------------

    /**
     * One (name, sql) pair — a constraint name paired with the validation SELECT that returns
     * violations. Mirrors the Python {@code (str, str)} tuple.
     */
    public record NamedSql(String name, String sql) {}

    /**
     * Options for {@link #emitValidation} and {@link #emitValidationUnion}.
     *
     * @param schema                    postgres schema (default {@code "knot_data"})
     * @param layer                     which view to validate against (default {@link Layer#RESOLVED})
     * @param scopeToSourceIdentifiers  when non-null, restrict each SELECT to canonical_ids
     *                                  touched by the given (source_name → source_identifiers)
     *                                  batch; {@code null} means full-table validation
     * @param includeBuiltins           when {@code true}, also emit structural invariants derived
     *                                  from the spec shape (FK-orphan + required-null checks)
     */
    public record ConstraintsOptions(
            String schema,
            Layer layer,
            Map<String, List<String>> scopeToSourceIdentifiers,
            boolean includeBuiltins) {

        /** Python kwarg defaults. */
        public static ConstraintsOptions defaults() {
            return new ConstraintsOptions("knot_data", Layer.RESOLVED, null, true);
        }
    }

    // -------------------------------------------------------------------------
    // Built-in constraints
    // -------------------------------------------------------------------------

    /**
     * Yield invariants knot ships automatically from the spec's shape.
     *
     * <p>Two families, both prefixed {@code _builtin_}:
     * <ol>
     *   <li>{@code _builtin_fk_orphan_<Class>_<slot>} — every ClassRef slot. Body:
     *       {@code slot.isNull() | slot.targetExists()}. Severity ERROR.</li>
     *   <li>{@code _builtin_required_null_<Class>_<slot>} — every required non-identifier slot.
     *       Body: {@code slot.isNotNull()}. Severity WARNING.</li>
     * </ol>
     *
     * <p>Virtual classes are skipped — built-ins only apply to concrete {@link OntologyClass}.
     */
    public static List<Constraint> builtinConstraints(Spec spec) {
        var out = new ArrayList<Constraint>();
        for (var raw : spec.classes().values()) {
            if (!(raw instanceof OntologyClass cls) || cls.kind() != ClassKind.CONCRETE) {
                continue;
            }
            for (Slot slot : cls.effectiveSlots()) {
                if (slot.type() instanceof ClassRef cr) {
                    var fkRef = new FkRef(cls.name(), slot.name(), cr.target().name());
                    var isNull = new IsNull(fkRef, false);
                    var targetExists = fkRef.targetExists();
                    var body = isNull.or_(targetExists);
                    String msg = cls.name() + "." + slot.name() + " → "
                            + cr.target().name() + ": FK value does not match any canonical_id";
                    out.add(new Constraint(
                            "_builtin_fk_orphan_" + cls.name() + "_" + slot.name(),
                            cls,
                            body,
                            Severity.ERROR,
                            msg));
                }
                if (slot.required() && !slot.identifier()) {
                    var slotRef = cls.col().get(slot.name());
                    var body = slotRef.isNotNull();
                    String msg = "required slot " + cls.name() + "." + slot.name()
                            + " resolved to NULL (no source provided a non-null value)";
                    out.add(new Constraint(
                            "_builtin_required_null_" + cls.name() + "_" + slot.name(),
                            cls,
                            body,
                            Severity.WARNING,
                            msg));
                }
            }
        }
        return out;
    }

    // -------------------------------------------------------------------------
    // Primary emitters
    // -------------------------------------------------------------------------

    /**
     * Return {@code (constraint_name, validation_sql)} pairs — one per constraint.
     *
     * <p>Each SQL returns zero rows when the constraint holds and one row per violating
     * canonical_id otherwise.
     *
     * <p>Mirrors the Python {@code emit_validation(spec, *, schema, layer,
     * scope_to_source_identifiers, include_builtins)}.
     *
     * <p>This overload is called by {@link Spec#emitValidation(Map, boolean)}.
     */
    public static List<Spec.ConstraintEntry> emitValidation(
            Spec spec,
            Map<String, List<String>> scopeToSourceIdentifiers,
            boolean includeBuiltins) {
        var opts = new ConstraintsOptions(
                spec.schema(),
                Layer.RESOLVED,
                scopeToSourceIdentifiers,
                includeBuiltins);
        var pairs = emitValidation(spec, opts);
        var out = new ArrayList<Spec.ConstraintEntry>(pairs.size());
        for (var p : pairs) {
            out.add(new Spec.ConstraintEntry(p.name(), p.sql()));
        }
        return out;
    }

    /**
     * Return {@link NamedSql} pairs — one per constraint, using explicit {@link ConstraintsOptions}.
     */
    public static List<NamedSql> emitValidation(Spec spec, ConstraintsOptions opts) {
        String scopeSql = buildScopeSql(opts.scopeToSourceIdentifiers());

        var all = new ArrayList<Constraint>(spec.constraints());
        if (opts.includeBuiltins()) {
            all.addAll(builtinConstraints(spec));
        }

        var out = new ArrayList<NamedSql>(all.size());
        for (var c : all) {
            out.add(new NamedSql(c.name(), buildSql(c, opts.schema(), opts.layer(), scopeSql)));
        }
        return out;
    }

    /**
     * Return a single {@code UNION ALL} of every constraint's validation SELECT, or {@code null}
     * when the spec has no constraints.
     *
     * <p>Mirrors Python {@code emit_validation_union}.
     */
    public static String emitValidationUnion(Spec spec, ConstraintsOptions opts) {
        var parts = emitValidation(spec, opts);
        if (parts.isEmpty()) {
            return null;
        }
        var sb = new StringBuilder();
        for (int i = 0; i < parts.size(); i++) {
            if (i > 0) sb.append("\nUNION ALL\n");
            // Strip trailing semicolon from each individual SELECT.
            String sql = parts.get(i).sql();
            if (sql.endsWith(";")) {
                sql = sql.substring(0, sql.length() - 1);
            }
            sb.append(sql);
        }
        sb.append(";");
        return sb.toString();
    }

    // -------------------------------------------------------------------------
    // Private helpers
    // -------------------------------------------------------------------------

    private static String escapeLiteral(String s) {
        return s.replace("'", "''");
    }

    /**
     * Build the {@code (source_name, source_identifier) IN (…)} pairs SQL fragment, or
     * {@code null} when {@code scope} is null or empty.
     */
    private static String buildScopeSql(Map<String, List<String>> scope) {
        if (scope == null || scope.isEmpty()) {
            return null;
        }
        var pairs = new ArrayList<String>();
        for (var entry : scope.entrySet()) {
            String srcLit = "'" + escapeLiteral(entry.getKey()) + "'";
            for (String ident : entry.getValue()) {
                String identLit = "'" + escapeLiteral(ident) + "'";
                pairs.add("(" + srcLit + ", " + identLit + ")");
            }
        }
        if (pairs.isEmpty()) {
            return null;
        }
        return String.join(", ", pairs);
    }

    /** Build one validation SELECT for constraint {@code c}. */
    private static String buildSql(Constraint c, String schema, Layer layer, String scopeSql) {
        OntologyClass primary = c.primary();
        Slot identifier = primary.identifierSlot();
        String table = schema + "." + primary.name().toLowerCase() + layer.suffix();
        String bindingsTable = schema + "." + primary.name().toLowerCase() + "_bindings";
        String messageLiteral = c.message() != null
                ? "'" + escapeLiteral(c.message()) + "'"
                : "NULL";

        String bodySql = ExprCompiler.compileSql(c.body(), schema, layer, primary.name());

        String scopeClause = "";
        if (scopeSql != null) {
            scopeClause = "\n  AND " + table + "." + identifier.name() + " IN (\n"
                    + "    SELECT DISTINCT canonical_id FROM " + bindingsTable + "\n"
                    + "    WHERE (source_name, source_identifier) IN (" + scopeSql + ")\n"
                    + "      AND canonical_id IS NOT NULL\n"
                    + "  )";
        }

        return "SELECT\n"
                + "    '" + escapeLiteral(c.name()) + "' AS rule_id,\n"
                + "    '" + escapeLiteral(primary.name()) + "' AS class_name,\n"
                + "    '" + escapeLiteral(c.severity().value()) + "' AS severity,\n"
                + "    " + messageLiteral + " AS message,\n"
                + "    " + identifier.name() + " AS offending_pk\n"
                + "FROM " + table + "\n"
                + "WHERE NOT (" + bodySql + ")" + scopeClause + ";";
    }
}
